File: python/src/azoth/eos/reference/ps_flash.py
"""``eos.ps_flash`` - the temperature a mixture reaches at a given pressure and entropy.

Spec: ``specs/models/eos/ps_flash.toml``

An outer bisection on temperature, for an **isentropic** unit operation - a compressor,
an expander, a turbine or a nozzle, assumed ideal - whose outlet temperature is not
known because entropy is conserved and temperature is not. The bracket, the loop, the
phase branch and the warning handling live in
:mod:`azoth.eos.reference._flash_property`, shared with ``eos.ph_flash``.

Each phase's entropy is taken at *that phase's own composition*, so the weighted sum
carries the entropy of mixing; summing the phases at the feed composition instead would
conserve entropy across a phase change, which is wrong.
"""

from __future__ import annotations

from typing import Any

from azoth import _models_gen
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import PsFlashResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.mixture import Mixture
from azoth.eos.reference._flash_property import Property, distinct_warnings, property_at
from azoth.eos.reference._flash_property import solve_temperature as _solve
from azoth.eos.reference.molar_enthalpy_entropy import IdealGasModel

MODEL_ID = "eos.ps_flash"

#: Which property this model inverts. The one thing it does not share with `ph_flash`.
WHICH: Property = "s"


def entropy_at(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    t_si: float,
    p_si: float,
    z: list[float],
) -> tuple[float, dict[str, Any]]:
    """The molar entropy of a mixture at a temperature and pressure, and its split.

    The composition the flash settles on is the equilibrium one, so this is the entropy
    of the *feed* at that state - which is what makes it comparable with the entropy a
    caller says the stream arrived with.
    """
    return property_at(mixture, ideal_gas, t_si, p_si, z, WHICH)


def solve_temperature(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    p_si: float,
    s_si: float,
    z: list[float],
    algorithm: dict[str, Any],
) -> dict[str, Any]:
    """Invert ``S(T, P) = s_si`` for ``T`` by bracketing, then bisecting."""
    return _solve(mixture, ideal_gas, p_si, s_si, z, WHICH, algorithm)


def ps_flash(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    P: Q,
    S: Q,
    z: list[float],
) -> PsFlashResult:
    """The temperature at which a mixture has a given molar entropy at a pressure.

    ``S`` is a *difference* from the datum ``ideal_gas`` carries, not an absolute
    quantity - the same caveat ``eos.ph_flash`` and ``eos.molar_enthalpy_entropy``
    carry, and for the same reason.

    Raises:
        InvalidInputError: if the composition or any ideal-gas vector is the wrong
            length, or if ``z`` is not a composition.
        OutOfRangeError: if ``P`` is not positive, or a range check on the answer fails.
        SolverNotConvergedError: if no temperature on the bracket covers the requested
            entropy, or if the bisection reaches its cap.
    """
    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    p_si = input_to_si(spec, "P", P)
    s_si = input_to_si(spec, "S", S)
    apply_checks(checks.on_input, {"P": p_si, "S": s_si}.get, warnings)

    solved = solve_temperature(mixture, ideal_gas, p_si, s_si, list(z), spec["algorithm"])
    warnings.extend(solved["warnings"])
    flash = solved["state"]["flash"]

    apply_checks(
        checks.derived,
        lambda name: solved["T"] if name == "T" else None,
        warnings,
    )

    return PsFlashResult(
        T=from_si(solved["T"], "K"),
        beta=flash.beta,
        x=tuple(flash.x),
        y=tuple(flash.y),
        k=tuple(flash.k),
        phase=flash.phase,
        z_liquid=flash.z_liquid,
        z_vapour=flash.z_vapour,
        iterations=solved["iterations"],
        residual=solved["residual"],
        warnings=distinct_warnings(warnings),
    )

File: crates/azoth-eos/src/ps_flash.rs
//! `eos.ps_flash` - the temperature a mixture reaches at a given pressure and entropy.
//!
//! Spec: `specs/models/eos/ps_flash.yaml`
//!
//! An outer bisection on temperature, for an **isentropic** unit operation - a
//! compressor, an expander, a turbine or a nozzle, assumed ideal - whose outlet
//! temperature is not known because entropy is conserved and temperature is not. The
//! bracket, the loop, the phase branch and the warning handling live in
//! [`crate::flash_property`], shared with `eos.ph_flash`.
//!
//! Each phase's entropy is taken at *that phase's own composition*, so the weighted sum
//! carries the entropy of mixing; summing the phases at the feed composition instead
//! would conserve entropy across a phase change, which is wrong.

use azoth_core::units::{MolarHeatCapacity, Pressure, ThermodynamicTemperature, kelvins};
use azoth_core::{Result, apply_checks};

use crate::algorithm_of;
use crate::flash_property::{Property, property_at, solve_temperature};
use crate::mixture::Mixture;
use crate::model_gen;
use crate::molar_enthalpy_entropy::IdealGasModel;
use crate::results::{PsFlashResult, PtFlashResult};

/// The molar entropy of a mixture at a temperature and pressure, and its split.
///
/// # Errors
/// Propagates whatever [`crate::pt_flash`] and [`crate::molar_enthalpy_entropy`] raise.
pub fn entropy_at(
    mixture: &Mixture,
    ideal_gas: &IdealGasModel,
    t: ThermodynamicTemperature,
    p: Pressure,
    z: &[f64],
) -> Result<(f64, PtFlashResult)> {
    property_at(mixture, ideal_gas, t, p, z, Property::Entropy)
}

/// The temperature at which a mixture has a given molar entropy at a pressure.
///
/// `s` is a *difference* from the datum `ideal_gas` carries, not an absolute quantity -
/// the same caveat `eos.ph_flash` and `eos.molar_enthalpy_entropy` carry.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `p` is not positive, or a range check on the answer
///   fails.
/// * [`azoth_core::AzothError::InvalidInput`] if the spec declares no bracket, which would be a
///   generator bug rather than a caller's.
/// * [`azoth_core::AzothError::SolverNotConverged`] if no temperature on the bracket covers the
///   requested entropy, or if the bisection reaches its cap.
///
/// # Example
/// ```
/// use azoth_core::units::{joules_per_mole_kelvin, pascals, kelvins};
/// use azoth_eos::mixture::{Component, Mixture};
/// use azoth_eos::molar_enthalpy_entropy::IdealGasModel;
/// use azoth_eos::ps_flash;
///
/// let mixture = Mixture::new(
///     vec![
///         Component::new(kelvins(190.56), pascals(4_599_000.0), 0.0115)?,
///         Component::new(kelvins(425.12), pascals(3_796_000.0), 0.2002)?,
///     ],
///     vec![0.0, 0.01289789, 0.01289789, 0.0],
/// )?;
/// let ideal_gas = IdealGasModel {
///     cp_a: vec![3.0, 5.0],
///     cp_b: vec![0.0, 0.0],
///     cp_c: vec![0.0, 0.0],
///     cp_d: vec![0.0, 0.0],
///     cp_e: vec![0.0, 0.0],
/// };
/// let r = ps_flash::ps_flash(
///     &mixture,
///     &ideal_gas,
///     pascals(2_000_000.0),
///     joules_per_mole_kelvin(-38.61276026788922),
///     &[0.6, 0.4],
/// )?;
/// assert!((r.temperature.value - 300.0).abs() < 1.0e-6);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
pub fn ps_flash(
    mixture: &Mixture,
    ideal_gas: &IdealGasModel,
    p: Pressure,
    s: MolarHeatCapacity,
    z: &[f64],
) -> Result<PsFlashResult> {
    let spec = &model_gen::PS_FLASH_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "P" => Some(p.value),
            "S" => Some(s.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let algorithm = algorithm_of(spec)?;
    let solved = solve_temperature(
        mixture,
        ideal_gas,
        p,
        s.value,
        z,
        Property::Entropy,
        algorithm,
    )?;

    apply_checks(
        spec.derived_checks(),
        |quantity| (quantity == "T").then_some(solved.temperature),
        &mut warnings,
    )?;

    Ok(PsFlashResult {
        temperature: kelvins(solved.temperature),
        beta: solved.flash.beta,
        x: solved.flash.x,
        y: solved.flash.y,
        k: solved.flash.k,
        phase: solved.flash.phase,
        z_liquid: solved.flash.z_liquid,
        z_vapour: solved.flash.z_vapour,
        iterations: solved.iterations,
        residual: solved.residual,
        warnings: solved.warnings,
    })
}
